### PR TITLE
Replace LANLstar/LANLmax with numpy-based implementation

### LINKED ISSUE #160: Handle ffnet dependency with binary install
Right now we hard-require ffnet in our setup.py declarations. That means even with a binary wheel built for Windows, doing a pip install on Windows will require a Fortran compiler for ffnet (and appropriate flags on the install.)

Long-term maybe we can contribute a wheel to ffnet. Short-term we can make LANLstar an "extra" (in setuptools terms). This means:
`pip install spacepy`

will not grab ffnet. (Of course, installing ffnet later would still allow LANLstar to work.) To make sure ffnet works you'd have to do:
`pip install spacepy[LANLstar]`

Now, long-term we also are talking about splitting LANLstar into a "kits" or something so this'll be a bit weird (how-to-get-LANLstar will change) but again, we're shooting for good short-term fixes right now.

Assuming we're okay with this, it should be possible to still hard-require ffnet on other platforms. (For that matter, we can also skip the "extras" thing and just tell people to install ffnet themselves if they want LANLstar on Windows.)

This is turning into multiple-choice, so:

1. Require a Fortran compiler to install a binary wheel on Windows (current)
2. Declare LANLstar an "extra" so you have to install SpacePy in a funny way to get ffnet and LANLstar working
    1. Do this on all platforms
    2. Do this just on Windows while having ffnet still base-required on other platforms
3. Do not require ffnet on Windows so people have to install ffnet explicitly on Windows, but will get it automatically on other platforms.

@drsteve ? @balarsen ?

### LINKED ISSUE #472: Consider vectorising LANLstar calculations
The current calculation of LANLstar values for a large number of input points/epochs is rather more inefficient than it needs to be. Specifically, the for-loop in the _LANLcommon, from LANLstar.py:

https://github.com/spacepy/spacepy/blob/76c5a35c688e7876292ce05f3bf833a793d77fd6/spacepy/LANLstar.py#L144-L148

The actual neural-network calculation that takes place under the ffnet hood in this loop is a simple series of matrix/vector operations, but there is a significant level of overhead involved in each call to ffnet. It is possible to extract the various weights and biases needed to manually calculate the ANN value from the ffnet pickle object, and then convert the for-loop above into a series of numpy vector operations. This results in significantly increased throughput for the LANLstar functions -- it also has the added benefit of removing the soft-dependency on ffnet+networkx.

I have a working proof-of-concept for this idea, which accurately reproduces the results from ffnet, however unfortunately it is currently written in MATLAB script (for various reasons it was easier for me to do this at the time in MATLAB than modify SpacePy to do this). It shouldn't be too arduous to convert it to work within SpacePy, but I guess the first question is whether this is even something that the SpacePy maintainers are interested in?

### LINKED ISSUE #572: ffnet error
There are long lines of error messages when installing SpacePy (pip install --upgrade spacepy) on my new Mac (M1, OS 12 Monterey).

I installed Anaconda so everything is fresh installed. In the middle of the messages, the following seems to be the main one:

  CCompilerOpt.cc_test_flags[1003] : testing failed

  CCompilerOpt.dist_test[576] : CCompilerOpt._dist_test_spawn[711] : Command (clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -fwrapv -O2 -Wall -fPIC -O2 -isystem /Users/hyominkim/opt/anaconda3/include -arch x86_64 -I/Users/hyominkim/opt/anaconda3/include -fPIC -O2 -isystem /Users/hyominkim/opt/anaconda3/include -arch x86_64 -I/Users/hyominkim/opt/anaconda3/include/python3.9 -c /Users/hyominkim/opt/anaconda3/lib/python3.9/site-packages/numpy/distutils/checks/cpu_avx512_knm.c -o /var/folders/tx/vdr71c196vq4jvqgpsm3wpxr0000gn/T/tmph0at3c6n/Users/hyominkim/opt/anaconda3/lib/python3.9/site-packages/numpy/distutils/checks/cpu_avx512_knm.o -MMD -MF /var/folders/tx/vdr71c196vq4jvqgpsm3wpxr0000gn/T/tmph0at3c6n/Users/hyominkim/opt/anaconda3/lib/python3.9/site-packages/numpy/distutils/checks/cpu_avx512_knm.o.d -msse -msse2 -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx -mf16c -mfma -mavx2 -mavx512f -mavx512cd -mavx512er -mavx512pf -Werror) failed with exit status 1 output ->

  /Users/hyominkim/opt/anaconda3/lib/python3.9/site-packages/numpy/distutils/checks/cpu_avx512_knm.c:9:9: error: implicit declaration of function '_mm512_4fmadd_ps' is invalid in C99 [-Werror,-Wimplicit-function-declaration]

      b = _mm512_4fmadd_ps(b, b, b, b, b, NULL);

          ^

  /Users/hyominkim/opt/anaconda3/lib/python3.9/site-packages/numpy/distutils/checks/cpu_avx512_knm.c:9:9: note: did you mean '_mm512_fmadd_ps'?

  /Library/Developer/CommandLineTools/usr/lib/clang/13.0.0/include/avx512fintrin.h:2716:1: note: '_mm512_fmadd_ps' declared here

  _mm512_fmadd_ps(__m512 __A, __m512 __B, __m512 __C)

  ^

  /Users/hyominkim/opt/anaconda3/lib/python3.9/site-packages/numpy/distutils/checks/cpu_avx512_knm.c:9:7: error: assigning to '__m512' (vector of 16 'float' values) from incompatible type 'int'

      b = _mm512_4fmadd_ps(b, b, b, b, b, NULL);

        ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  /Users/hyominkim/opt/anaconda3/lib/python3.9/site-packages/numpy/distutils/checks/cpu_avx512_knm.c:11:9: error: implicit declaration of function '_mm512_4dpwssd_epi32' is invalid in C99 [-Werror,-Wimplicit-function-declaration]

      a = _mm512_4dpwssd_epi32(a, a, a, a, a, NULL);

          ^

  /Users/hyominkim/opt/anaconda3/lib/python3.9/site-packages/numpy/distutils/checks/cpu_avx512_knm.c:11:9: note: did you mean '_mm512_dpwssd_epi32'?

  /Library/Developer/CommandLineTools/usr/lib/clang/13.0.0/include/avx512vnniintrin.h:68:1: note: '_mm512_dpwssd_epi32' declared here

  _mm512_dpwssd_epi32(__m512i __S, __m512i __A, __m512i __B)

  ^

  /Users/hyominkim/opt/anaconda3/lib/python3.9/site-packages/numpy/distutils/checks/cpu_avx512_knm.c:11:7: error: assigning to '__m512i' (vector of 8 'long long' values) from incompatible type 'int'

      a = _mm512_4dpwssd_epi32(a, a, a, a, a, NULL);

        ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  4 errors generated.

  

  CCompilerOpt.feature_test[1458] : testing failed

As well as these lines,

    ld: library not found for -lSystem

    collect2: error: ld returned 1 exit status

    error: Command "/Users/hyominkim/opt/anaconda3/bin/x86_64-apple-darwin13.4.0-gfortran -Wall -g -arch x86_64 -Wall -g -undefined dynamic_lookup -bundle build/temp.macosx-10.9-x86_64-3.9/build/src.macosx-10.9-x86_64-3.9/ffnet/fortran/_ffnetmodule.o build/temp.macosx-10.9-x86_64-3.9/build/src.macosx-10.9-x86_64-3.9/build/src.macosx-10.9-x86_64-3.9/ffnet/fortran/fortranobject.o build/temp.macosx-10.9-x86_64-3.9/ffnet/fortran/ffnet.o build/temp.macosx-10.9-x86_64-3.9/build/src.macosx-10.9-x86_64-3.9/ffnet/fortran/_ffnet-f2pywrappers.o -L/Users/hyominkim/opt/anaconda3/bin/../lib/gcc/x86_64-apple-darwin11.4.2/4.8.5 -L/Users/hyominkim/opt/anaconda3/bin/../lib/gcc/x86_64-apple-darwin11.4.2/4.8.5/../../.. -L/Users/hyominkim/opt/anaconda3/bin/../lib/gcc/x86_64-apple-darwin11.4.2/4.8.5/../../.. -lgfortran -o build/lib.macosx-10.9-x86_64-3.9/ffnet/fortran/_ffnet.cpython-39-darwin.so" failed with exit status 1

### PR DESCRIPTION
**Pull request by drsteve**
The series of LANLstar models were all built using `ffnet`, which brought dependencies on both `ffnet` and `networkx`. As `ffnet` is not easily installed on some systems (depending on compiler availability, version of setuptools, version of Python, etc.), this refactor drops the external dependencies and replaces with a pure numpy implementation. This follows and closes #472.

Other issues impacted by this PR:
- Closes #572 as no longer relevant
- Closes #160 
- We can decide whether we think #521 should be closed following this too

The highlights here are that:
1. Testing is expanded to cover the case of multiple times requested (old tests only covered LANLstar for a single time
2. The old `.net` files (pickles of `ffnet`/`networkx` objects) have been converted to much smaller JSON-headed ASCII files
3. Dependence on `ffnet` and `networkx` has been dropped. While the code could be sped up (I just got lost in array broadcasting and decided a working version was the goal), it's actually ~30% faster than the old implementation right now anyway.

### PR Checklist
- [x] Pull request has descriptive title
- [x] Pull request gives overview of changes
- [x] New code has inline comments where necessary
- [x] Any new modules, functions or classes have docstrings consistent with SpacePy style
- [x] Added an entry to release notes if fixing a significant bug or providing a new feature
- [x] New features and bug fixes should have unit tests
- [x] Relevant issues are linked to (e.g. `See issue #` or `Closes #`)

### CODE REVIEW

**Review by jtniehof — CHANGES_REQUESTED** [review]
Works, elegant, no problems. Piddly stuff:

- There's going to be a merge conflict with #614. I'm happy to do the rebase and fix.
- Related, the release notes have lines pushing 100 characters. PEP8 says 72 for documentation (but that's more for docstrings); numpydoc is also 75 for docstrings. The [Python style guide](https://devguide.python.org/documenting/#style-guide) is 80. I'm happy to rewrap while fixing the rebase conflicts. (And yes, long-term we need our own style guide so we can have these conversations once).
- The use of `__file__` in `_get_net_path` can cause problems with e.g. zip deployments and probably some executable deployment helper things too. That's how we were doing it before, so I suggest we *not* change at this time, but I should probably open another issue for converting all such usage to `pkgutil.get_data`.
- The `test_all.sh` script was still grabbing ffnet...again, I'm happy to just make that change myself.

So basically I'm just looking for "you fix that" or "I'll fix that" and then I'll (potentially make changes for you to review) and merge.

**Comment by drsteve** [comment]
I numbered your bullet points because counting is hard and I wanted to make it easier to respond...

1. Sure, happy for you to do this.
2. When we have this conversation later we can rewrap whatever we need to. If you feel the burning desire to wrap at numpydoc length or whatever then go ahead, but if I had my way the wrap for everything would happen at 99. Too short lines irk me, unless I'm on my phone. TL;DR: Wrap in whatever way will make you happy, or don't. Either works for me.
3. Point taken, but like you said it's not a change in how things were done. Happy for you to open an issue so we can fix smartly in future.
4. I might as well volunteer to do something useful and this sounds like it's an oversight on my end, so I'll fix that now. Once that's in I'll leave the rest to you and can look over any additional changes you make alongside the rebase.

**Review by jtniehof — APPROVED** [review]
Approved with these changes; I opened #618 for the data reading issue.
